### PR TITLE
Add survey publishing and server-side validations

### PR DESCRIPTION
**Pull request by doodlehead**
Adding server-side validations for survey publishing and closing

## Publishing
- You cannot answer a survey that is not published
- You cannot publish a closed survey
- You cannot add questions to a published survey

## Closing
- You cannot answer a closed survey
- You cannot add questions to a closed survey

These use cases are tested in `RestTest.java`

### CODE REVIEW

File: src/main/java/com/example/demo/controller/QuestionsController.java [hunk 1]
```diff
@@ -38,8 +42,12 @@ public Question get(@PathVariable int id) {
 	 * @return
 	 */
 	@PostMapping("/question")
-	public Question create(@RequestBody Question question) {
-		return questionsRepository.save(question); 
+	public ResponseEntity<Question> create(@RequestBody Question question) {
+		Survey survey = question.getSurvey();
+		if (survey.isClosed() || !survey.isPublished()) {
```
Comment (Binyamin-Brion) [inline src/main/java/com/example/demo/controller/QuestionsController.java]: Should this be if the survey is closed or if the survey _is_ published?

File: src/test/java/com/example/demo/QuestionTest.java [hunk 2]
```diff
@@ -37,31 +37,6 @@ void testQuestionType() {
 		assertEquals(range.getType(), QuestionType.RANGE);
 		assertEquals(mc.getType(), QuestionType.MC);
 	}
-
-	@Test
-	void controllerTest() {
```
Comment (Binyamin-Brion) [inline src/test/java/com/example/demo/QuestionTest.java:42]: Why is this test removed?

File: src/test/java/com/example/demo/RestTest.java [hunk 3]
```diff
@@ -52,4 +53,71 @@ public void testAuthentication() throws Exception {
                 .andDo(print())
                 .andExpect(status().isOk());
     }
+
+    @Test
+    public void testSurvey() throws Exception {
+        // Create user
+        this.mockMvc.perform(post("/user")
```
Comment (Binyamin-Brion) [inline src/test/java/com/example/demo/RestTest.java:60]: Does this test requires logging in for the user?

File: src/main/java/com/example/demo/controller/SurveyController.java [hunk 4]
```diff
@@ -50,5 +52,16 @@ public Survey closeSurvey(@PathVariable int id) {
 		survey.setClosed(true);
 		return repository.save(survey);
 	}
+
+	@PostMapping("/survey/{id}/publish")
+	public ResponseEntity<Survey> publishSurvey(@PathVariable int id) {
```
Comment (Binyamin-Brion) [inline src/main/java/com/example/demo/controller/SurveyController.java:57]: Why does this return a survey?

File: src/main/java/com/example/demo/controller/SurveyController.java [hunk 5]
```diff
@@ -50,5 +52,16 @@ public Survey closeSurvey(@PathVariable int id) {
...
+		Survey survey = repository.findById(id).get();
+
+		if (survey.isClosed()) {
```
Comment (Binyamin-Brion) [inline src/main/java/com/example/demo/controller/SurveyController.java:60]: Should there extra check to see if survey is already published?

File: src/main/java/com/example/demo/controller/SurveyController.java (hunk 4 above)
Comment (doodlehead) [reply src/main/java/com/example/demo/controller/SurveyController.java:57]: Client-side might want to use this info to update

File: src/test/java/com/example/demo/RestTest.java (hunk 3 above)
Comment (doodlehead) [reply src/test/java/com/example/demo/RestTest.java:60]: Yes

File: src/test/java/com/example/demo/QuestionTest.java (hunk 2 above)
Comment (doodlehead) [reply src/test/java/com/example/demo/QuestionTest.java:42]: Breaks other tests

File: src/main/java/com/example/demo/controller/SurveyController.java (hunk 5 above)
Comment (doodlehead) [reply src/main/java/com/example/demo/controller/SurveyController.java:60]: Not super critical, publishing a published survey doesn't break anything

**Review by Binyamin-Brion — APPROVED** [review]

**Pull request merged**